rpcc/conn.go

```diff
@@ -1,536 +1,571 @@
-// package compress/flate for a description of compression levels.
-func (c *Conn) SetCompressionLevel(level int) error {
-	if c.compressionLevel == nil {
-		return errors.New("rpcc: compression is not enabled for Conn")
-	}
-	return c.compressionLevel(level)
-}
-
-// Close closes the connection.
-func (c *Conn) Close() error {
-	return c.close(nil)
-}
-
-// Debugging, enabled in tests.
-var enableDebug = false
+package rpcc
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+var (
+	// ErrConnClosing indicates that the operation is illegal because
+	// the connection is closing.
+	ErrConnClosing = errors.New("rpcc: the connection is closing")
+)
+
+const (
+	defaultWriteBufferSize = 4096
+)
+
+// DialOption represents a dial option passed to Dial.
+type DialOption func(*dialOptions)
+
+// WithCodec returns a DialOption that sets the codec responsible for
+// encoding and decoding requests and responses onto the connection.
+// This option overrides the default json codec.
+func WithCodec(f func(conn io.ReadWriter) Codec) DialOption {
+	return func(o *dialOptions) {
+		o.codec = f
+	}
+}
+
+// WithDialer returns a DialOption that sets the dialer for the underlying
+// connection. It can be used to replace the WebSocket library used by this
+// package or to communicate over a different protocol.
+//
+// This option overrides the default WebSocket dialer and all of:
+// WithCompression, WithTLSConfig and WithWriteBufferSize become no-op.
+func WithDialer(f func(ctx context.Context, addr string) (io.ReadWriteCloser, error)) DialOption {
+	return func(o *dialOptions) {
+		o.dialer = f
+	}
+}
+
+// WithWriteBufferSize returns a DialOption that sets the size of the write
+// buffer for the underlying websocket connection. Messages larger than this
+// size are fragmented according to the websocket specification.
+//
+// The maximum buffer size for recent versions of Chrome is 104857586 (~100MB),
+// for older versions a maximum of 1048562 (~1MB) can be used. This is because
+// Chrome does not support websocket fragmentation.
+func WithWriteBufferSize(n int) DialOption {
+	return func(o *dialOptions) {
+		o.wsDialer.WriteBufferSize = n
+	}
+}
+
+// WithCompression returns a DialOption that enables compression for the
+// underlying websocket connection. Use SetCompressionLevel on Conn to
+// change the default compression level for subsequent writes.
+func WithCompression() DialOption {
+	return func(o *dialOptions) {
+		o.wsDialer.EnableCompression = true
+	}
+}
+
+// WithTLSClientConfig specifies the TLS configuration to use with tls.Client.
+func WithTLSClientConfig(c *tls.Config) DialOption {
+	return func(o *dialOptions) {
+		o.wsDialer.TLSClientConfig = c
+	}
+}
+
+// WithTimeout specifies the timeout for every read and write operation of net.Conn.
+func WithTimeout(timeout time.Duration) DialOption {

```

Comment:
If we're to include this, we'd need to split this into `WithReadTimeout` and `WithWriteTimeout` or simply a `WithReadAndWriteTimeout`. Reason being is this could be confused with dial timeout (i.e. used only during dial).

Reply:
Yeah, that makes sense.